File: mobile/src/main/java/com/mattrubacky/monet2/deserialized/splatoon/Schedules.java
package com.mattrubacky.monet2.deserialized.splatoon;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;

/**
 * Created by mattr on 9/23/2017.
 * The Schedules for all gamemodes
 */


public class Schedules implements Parcelable {
    public Schedules(){

    }
    @SerializedName("regular")
    public ArrayList<TimePeriod> regular;
    @SerializedName("gachi")
    public ArrayList<TimePeriod> ranked;
    @SerializedName("league")
    public ArrayList<TimePeriod> league;

    //splatfest is not in the Splatnet API, this is populated client side
    @SerializedName("fes")
    public ArrayList<TimePeriod> splatfest;

    //Dequeue the expired schedules
    public void dequeue(){

        if(splatfest!=null&&splatfest.size()>0&&splatfest.get(0).start<regular.get(0).start){
            splatfest.remove(0);
        }
        if((splatfest!=null&&splatfest.get(0).start>regular.get(0).start)) {
            regular.remove(0);
            ranked.remove(0);
            league.remove(0);
        }
    }
    //Move Splatfest TimePeriods to fes, and remove other TimePeriods for that time
    public void setSplatfest(Splatfest splatfest){
        long start = splatfest.times.start;
        long end = splatfest.times.end;
        this.splatfest = new ArrayList<>();
        for(int i=0;i<regular.size();i++){
            while(regular.size()>i&&(regular.get(i).end>start)&&(regular.get(i).start<=end)){
                this.splatfest.add(regular.get(i));
                regular.remove(i);
                ranked.remove(i);
                league.remove(i);
            }
        }
    }

    protected Schedules(Parcel in) {
        regular = in.createTypedArrayList(TimePeriod.CREATOR);
        ranked = in.createTypedArrayList(TimePeriod.CREATOR);
        league = in.createTypedArrayList(TimePeriod.CREATOR);
    }

    public static final Creator<Schedules> CREATOR = new Creator<Schedules>() {
        @Override
        public Schedules createFromParcel(Parcel in) {
            return new Schedules(in);
        }

        @Override
        public Schedules[] newArray(int size) {
            return new Schedules[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeTypedList(regular);
        dest.writeTypedList(ranked);
        dest.writeTypedList(league);
    }
}

File: mobile/src/main/java/com/mattrubacky/monet2/data/deserialized/splatoon/parsley/Schedules.kt
package com.mattrubacky.monet2.data.deserialized.splatoon.parsley

import android.os.Parcel
import android.os.Parcelable
import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import com.google.gson.annotations.SerializedName
import com.mattrubacky.monet2.backend.api.splatnet.SplatnetParsley
import com.mattrubacky.monet2.backend.getCurrentTime
import com.mattrubacky.monet2.backend.insertStages
import com.mattrubacky.monet2.data.deserialized.splatoon.Splatfest
import com.mattrubacky.monet2.data.deserialized_entities.Stage
import com.mattrubacky.monet2.data.deserialized_entities.TimePeriod
import com.mattrubacky.monet2.data.parsley.Sprig
import com.mattrubacky.monet2.data.rooms.SplatnetDatabase
import retrofit2.Response
import kotlin.collections.ArrayList

class Schedules() : Sprig<SplatnetParsley,SplatnetDatabase>,Parcelable{

    var needsUpdate: Boolean = true
        private set
    private var stages: List<Stage> = ArrayList()

    @SerializedName("regular")
    var regular: ArrayList<TimePeriod> = ArrayList()
    @SerializedName("gachi")
    var ranked: ArrayList<TimePeriod> = ArrayList()
    @SerializedName("league")
    var league: ArrayList<TimePeriod> = ArrayList()

    //splatfest is not in the Splatnet API, this is populated client side
    @SerializedName("fes")
    var splatfest: ArrayList<TimePeriod> = ArrayList()

    private fun setSplatfest(splatfest: Splatfest) {
        val start = splatfest.times.start
        val end = splatfest.times.end
        this.splatfest = java.util.ArrayList()
        for (i in regular.indices) {
            while (regular.size > i && regular[i].end > start && regular[i].start <= end) {
                this.splatfest.add(regular[i])
                regular.removeAt(i)
                ranked.removeAt(i)
                league.removeAt(i)
            }
        }
    }


    override suspend fun getFromSource(source: SplatnetParsley) {
        if(needsUpdate) {
            val schedules: Response<Schedules> = source.getSchedules().execute()
            val splatfest: Response<CurrentSplatfest> = source.getActiveSplatfests().execute()
            if (schedules.isSuccessful) {
                val schedule = schedules.body()!!
                regular = schedule.regular
                ranked = schedule.ranked
                league = schedule.league
                if (splatfest.isSuccessful && splatfest.body()!!.splatfests.isNotEmpty()) {
                    setSplatfest(splatfest.body()!!.splatfests[0])
                }
            }
            needsUpdate = false
        }
    }

    override suspend fun checkStore(store: SplatnetDatabase) {
        val old = store.timePeriodDao.selectOld(getCurrentTime())
        for(timePeriod in old){
            store.timePeriodDao.delete(timePeriod)
        }
        needsUpdate = old.isNotEmpty()
    }

    override suspend fun putInStore(store: SplatnetDatabase) {
        for(timePeriod in regular){
            store.timePeriodDao.insertTimePeriod(timePeriod,store.stageDao)
        }
        for(timePeriod in ranked){
            store.timePeriodDao.insertTimePeriod(timePeriod,store.stageDao)
        }
        for(timePeriod in league){
            store.timePeriodDao.insertTimePeriod(timePeriod,store.stageDao)
        }
        for(timePeriod in splatfest){
            store.timePeriodDao.insertTimePeriod(timePeriod,store.stageDao)
        }
    }

    override fun getLiveData(store: SplatnetDatabase): LiveData<Sprig<SplatnetParsley,SplatnetDatabase>> {
        val schedules:MediatorLiveData<Sprig<SplatnetParsley,SplatnetDatabase>> = MediatorLiveData()

        schedules.addSource(store.stageDao.selectAllLive()) { result: List<Stage>? ->
            val updatedValue = schedules.value!! as Schedules
            updatedValue.stages = result ?: ArrayList()
            if (updatedValue.stages.isNotEmpty()){
                if (!updatedValue.regular.isNullOrEmpty()) {
                    updatedValue.regular = insertStages(updatedValue.regular, updatedValue.stages)
                }
                if(!updatedValue.ranked.isNullOrEmpty()){
                    updatedValue.ranked = insertStages(updatedValue.ranked, updatedValue.stages)
                }
                if(!updatedValue.league.isNullOrEmpty()){
                    updatedValue.league = insertStages(updatedValue.league, updatedValue.stages)
                }
                if(!updatedValue.splatfest.isNullOrEmpty()){
                    updatedValue.splatfest = insertStages(updatedValue.splatfest, updatedValue.stages)
                }
            }
            schedules.setValue(updatedValue)
        }

        schedules.addSource(store.timePeriodDao.selectRegularLive(getCurrentTime())){ result: List<TimePeriod>? ->
            val updatedValue = schedules.value!! as Schedules
            updatedValue.regular = ArrayList(result!!)
            if(updatedValue.stages.isNotEmpty()&&updatedValue.regular.isNotEmpty()) {
                updatedValue.regular = insertStages(updatedValue.regular, updatedValue.stages)
            }
            schedules.setValue(updatedValue)
        }
        schedules.addSource(store.timePeriodDao.selectGachiLive(getCurrentTime())){ result: List<TimePeriod>? ->
            val updatedValue =schedules.value!! as Schedules
            updatedValue.ranked = ArrayList(result!!)
            if(updatedValue.stages.isNotEmpty()&&updatedValue.ranked.isNotEmpty()) {
                updatedValue.ranked = insertStages(updatedValue.ranked, updatedValue.stages)
            }
            schedules.setValue(updatedValue)
        }
        schedules.addSource(store.timePeriodDao.selectLeagueLive(getCurrentTime())){ result: List<TimePeriod>? ->
            val updatedValue = schedules.value!! as Schedules
            updatedValue.league = ArrayList(result!!)
            if(updatedValue.stages.isNotEmpty()&&updatedValue.league.isNotEmpty()) {
                updatedValue.league = insertStages(updatedValue.league, updatedValue.stages)
            }
            schedules.setValue(updatedValue)
        }
        schedules.addSource(store.timePeriodDao.selectFestivalLive(getCurrentTime())){ result: List<TimePeriod>? ->
            val updatedValue = schedules.value!! as Schedules
            updatedValue.splatfest = ArrayList(result!!)
            if(updatedValue.stages.isNotEmpty()&&updatedValue.splatfest.isNotEmpty()) {
                updatedValue.splatfest = insertStages(updatedValue.splatfest, updatedValue.stages)
            }
            schedules.setValue(updatedValue)
        }

        return schedules
    }

    constructor(parcel: Parcel) : this() {
        regular = parcel.createTypedArrayList(TimePeriod.CREATOR)!!
        ranked = parcel.createTypedArrayList(TimePeriod.CREATOR)!!
        league = parcel.createTypedArrayList(TimePeriod.CREATOR)!!
        splatfest = parcel.createTypedArrayList(TimePeriod.CREATOR)!!
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeTypedList(regular)
        parcel.writeTypedList(ranked)
        parcel.writeTypedList(league)
        parcel.writeTypedList(splatfest)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<Schedules> {
        override fun createFromParcel(parcel: Parcel): Schedules {
            return Schedules(parcel)
        }

        override fun newArray(size: Int): Array<Schedules?> {
            return arrayOfNulls(size)
        }
    }
}